nevergrad/functions/powersystems/core.py

```diff
@@ -0,0 +1,193 @@
+        for i in range(layers-2):
+            self.layers += [np.random.rand(layer_width, layer_width)]
+        self.layers += [np.random.rand(layer_width, output_size)]
+        assert len(self.layers) == layers
+
+    def GetParamNumbers(self):
+        return sum([np.prod(l.shape) for l in self.layers])
+
+    def SetParams(self, ww):
+        w = [w for w in ww]
+        assert(len(w) == self.GetParamNumbers())
+        for l in self.layers:
+            s = np.prod(l.shape)
+            l = np.reshape(np.array(w[:s]), l.shape)
+            w = w[s:]
+
+    def GetOutput(self, inp):
+        output = np.array(inp).reshape(1, len(inp))
+        for l in [self.layers[i] for i in range(len(self.layers) - 1)]:
+            output = np.tanh(np.matmul(output, l))
+        return np.matmul(output, self.layers[-1])
+
+
+
+
+class PowerSystemVariable(Variable):
+
+    def __init__(self, num_stocks:int, depth: int, width: int, dimension: int) -> None:
+        
+        # Simple instrumentation: just the number of params.
+        super().__init__()
+        self.num_stocks = num_stocks
+        self.depth = depth
+        self.width = width
+        self._specs.update(dimension=dimension)
+
+    # pylint: disable=unused-argument
+    def _data_to_arguments(self, data: np.ndarray, deterministic: bool = True) -> Any:
+        assert len(data) == self.dimension
+        return data, {}
+
+    def __repr__(self) -> str:
+        return "PowerSystems" + str(self.num_stocks) + "stocks_" + str(self.depth) + "layers_" + str(self.width) + "neurons"
+
+
+class PowerSystem(inst.InstrumentedFunction):
+    """
+    Parameters
+    ----------
+    num_stocks: number of stocks to be managed
+    depth: number of layers in the neural networks
+    width: number of neurons per hidden layer
+    """
+
+    def __init__(self, num_stocks: int = 3, depth: int = 3, width: int = 3) -> None:
+        dam_managers: List[Any] = []
+        # Number of stocks (dams).
+        N = num_stocks
+        # Parameters describing the problem.
+        year_to_day_ratio = 2. * N  # Ratio between variation of consumption in the year and variation of consumption in the day
+        constant_to_year_ratio = N * 2.
+        back_to_normal = 0.03  # How much of the gap with normal is cancelled at each iteration.
+        consumption_noise = 1.
+        num_thermal_plants = 7
+        number_of_years = 1
+        
+        average_consumption = constant_to_year_ratio * year_to_day_ratio
+        thermal_power_capacity = [c * average_consumption for c in list(np.random.rand(num_thermal_plants))]
+        thermal_power_prices = [c for c in list(np.random.rand(num_thermal_plants))]
+        
+        for i in range(N):
+            dam_managers += [Agent(6 + N + 2*num_thermal_plants, 1)]
+        dimension = sum([a.GetParamNumbers() for a in dam_managers])
+        self._dimension = dimension
+
+        def _simulate_power_system(input_x: np.ndarray):
```

Comment:
please always avoid functions inside functions, they make pickling complicated.
Also, it will help structure the code, a function of more than about 30 line start being a bit complicated to read

Reply:
Removed from inside the function!